src/wp-includes/media.php

```diff
@@ -5753,11 +5758,44 @@ function wp_get_loading_optimization_attributes( $tag_name, $attr, $context ) {
 	 * If the element is in the viewport (`true`), potentially add
 	 * `fetchpriority` with a value of "high". Otherwise, i.e. if the element
 	 * is not not in the viewport (`false`) or it is unknown (`null`), add
-	 * `loading` with a value of "lazy".
+	 * `loading` with a value of "lazy" and set `decoding` with the value of
+	 * "async".
 	 */
 	if ( $maybe_in_viewport ) {
 		$loading_attrs = wp_maybe_add_fetchpriority_high_attr( $loading_attrs, $tag_name, $attr );
 	} else {
+		// Only add `decoding` attribute if the feature is enabled.
+		if ( 'img' === $tag_name ) {
+			$decoding = array_key_exists( 'decoding', $attr ) ? $attr['decoding'] : 'async';
+
+			// This filter is applied to ensure backward-compatibility.
+			/** This action is documented in wp-includes/media.php */
+			$decoding = apply_filters( 'wp_img_tag_add_decoding_attr', $decoding, $tag_name, $context );
+
+			/**
+			 * Filters the `decoding` attribute value to add to an image. Default `async`.
+			 *
+			 * Returning a falsy value will omit the attribute.
+			 *
+			 * @since 6.4.0
+			 *
+			 * @param string|false|null $value    The `decoding` attribute value. Returning a falsy value
+			 *                                    will result in the attribute being omitted for the image.
+			 *                                    Otherwise, it may be: 'async' (default), 'sync', or 'auto'.
+			 * @param string            $tag_name The tag name of the element.
+			 * @param string            $context  Additional context about how the function was called
+			 *                                    or where the img tag is.
+			 *
+			 *
+			 */
+			$decoding = apply_filters( 'wp_decoding_value', $decoding, $tag_name, $context );
+
+			// Validate the `decoding` attribute according to the spec.
+			if ( ! empty( $decoding ) && in_array( $decoding, array( 'async', 'sync', 'auto' ), true ) ) {
+				$loading_attrs['decoding'] = $decoding;
+			}
+		}
+
```

Comment:
@felixarntz 

About your points:
- I agree. Let's keep the existing `wp_img_tag_add_decoding_attr` filter. It will require the deprecation of the $image attribute as we no longer have access to it, but we can add the `$attr` array, which contains the attributes of said tag;
- These next two points got me thinking a lot. But I still believe that this is where the logic should be living. This is also what the ticket [#58892](https://core.trac.wordpress.org/ticket/58892) requests. I could move away the `decoding` validation from `wp_get_loading_optimization_attributes` to `wp_get_loading_optimization_attributes` (where I forgot to push one last change), but keeping it part of the root function will benefit all the users that are looking to get the WordPress image optimization attributes, regardless of the use case. I'd be very happy to hop on a call with you to check this out more closely.

Reply:
@pereirinha I think there's still a misalignment here. Let me try to clarify:
* The `wp_img_tag_add_decoding_attr` filter was previously applied in `wp_img_tag_add_decoding_attr()`.
* The function that now replaces `wp_img_tag_add_decoding_attr()` is `wp_img_tag_add_loading_optimization_attrs()`, not `wp_get_loading_optimization_attributes()`. So the filter should be run in the first function, not the latter, otherwise this breaks the expectation that this filter is for any in-content images, which is problematic in terms of backward compatibility. Please review the `wp_img_tag_add_loading_attr` filter for how that was migrated from the `wp_img_tag_add_loading_attr()` function to `wp_img_tag_add_loading_optimization_attrs()` - we should follow the same approach for the filter here.
* The difference between `wp_img_tag_add_loading_optimization_attrs()` and `wp_get_loading_optimization_attributes()` is that the latter is the source of truth function that is applied for every image tag anywhere, while the first is only applied for in-content images. This may be a bit confusing and is only like that because initially there was no centralized function (`wp_get_loading_optimization_attributes()` was only introduced in WordPress 6.3).

I understand your desire to handle the logic centrally, but this filter historically has _not_ been an all-encompassing filter, and it also uses a naming pattern that we probably want to get rid of in the long term. So for backward compatibility reasons, the filter should only be run on image tags within the content blob, i.e. in `wp_img_tag_add_loading_optimization_attrs()`.